x/valset-pref/validator_set.go

```diff
@@ -119,72 +128,125 @@ func (k Keeper) DelegateToValidatorSet(ctx sdk.Context, delegatorAddr string, co
 			return err
 		}
 
-		// in the last valset iteration we dont calculate it from shares using decimals and trucation,
-		// we use whats remaining to get more accurate value
-		if len(existingSet.Preferences)-1 == i {
-			amountToUnDelegate = coin.Amount.Sub(totalUnDelAmt).ToDec().TruncateInt()
-		} else {
-			// Calculate the amount to undelegate based on the existing weights
-			amountToUnDelegate = val.Weight.Mul(tokenAmt).TruncateInt()
-			totalUnDelAmt = totalUnDelAmt.Add(amountToUnDelegate)
+		delegation, found := k.stakingKeeper.GetDelegation(ctx, delegator, valAddr)
+		if !found {
+			return fmt.Errorf("No delegation found for thie delegator %s to this validator %s\n", delegator, valAddr)
 		}
 
 		sharesAmt, err := validator.SharesFromTokens(amountToUnDelegate)
 		if err != nil {
 			return err
 		}
 
-		_, err = k.stakingKeeper.Undelegate(ctx, delegator, valAddr, sharesAmt) // this has to be shares amount
+		valShares := sharesAmt.Quo(delegation.Shares)
+		valSetRatio = append(valSetRatio, ValRatio{
+			ValAddr:       valAddr,
+			UndelegateAmt: amountToUnDelegate,
+			VRatio:        valShares,
+			DelegatedAmt:  delegation.Shares.TruncateInt(),
+			Weight:        val.Weight,
+		})
+	}
+
+	// Step 3
+	sort.Slice(valSetRatio, func(i, j int) bool {
+		return valSetRatio[i].VRatio.GT(valSetRatio[j].VRatio)
+	})
+
+	// Step 4
+	totalUnDelAmt, amountToUnDelegate := sdk.NewInt(0), sdk.NewInt(0)
+	if valSetRatio[0].VRatio.LTE(sdk.OneDec()) {
+		fmt.Println("All validators can undelegate normally, falling to happy path exit early")
+		for index, val := range valSetRatio {
+			_, validator, err := k.getValAddrAndVal(ctx, val.ValAddr.String())
+			if err != nil {
+				return err
+			}
+
+			// in the last valset iteration we dont calculate it from shares using decimals and trucation,
+			// we use whats remaining to get more accurate value
+			if len(existingSet.Preferences)-1 == index {
+				amountToUnDelegate = coin.Amount.Sub(totalUnDelAmt).ToDec().TruncateInt()
+			} else {
+				// Calculate the amount to undelegate based on the existing weights
+				amountToUnDelegate = val.UndelegateAmt
+				totalUnDelAmt = totalUnDelAmt.Add(amountToUnDelegate)
+			}
+
+			sharesAmt, err := validator.SharesFromTokens(amountToUnDelegate)
+			if err != nil {
+				return err
+			}
+
+			_, err = k.stakingKeeper.Undelegate(ctx, delegator, val.ValAddr, sharesAmt) // this has to be shares amount
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	// Step 5
+	// `targetRatio`: This is a threshold value that is used to decide how to unbond tokens from validators.
+	// It starts as 1 and is recalculated each time a validator is fully unbonded and removed from the unbonding process.
+	// By reducing the target ratio using the ratio of the removed validator, we adjust the proportions we are aiming for with the remaining validators.
+	targetRatio := sdk.OneDec()
+	amountRemaining := coin.Amount
+
+	fmt.Println("VRATIO PRE: ", valSetRatio)
```

Comment:
note: these Println are easier to debug/review. Will remove them once this PR is close to merge